azure-kusto-data/tests/aio/test_async_token_providers.py

```diff
@@ -313,3 +313,25 @@ async def test_cloud_mfa_on(self):
         assert context["authority"] == "https://login_endpoint/auth_test"
         assert context["client_id"] == "1234"
         assert provider._scopes == ["https://fakeurl.kustomfa.windows.net/.default"]
+
+    def test_async_lock(self):
+        """
+        This test makes sure that the lock inside of a TokenProvider, is created within the correct event loop.
+        Before this, the Lock was created once per class.
+        This meant that if someone created a new event loop, and created a provider in it, awaiting on the lock would cause an exception because it belongs to
+        a different loop.
+        Now the lock is instantiated for every class instance, avoiding this issue.
+        """
+
+        async def start():
+            provider = MockProvider(KUSTO_URI)
```

Comment:
So, noticed basic and callback providers are not async compatible.
a) why?
b) had callback been async you'd be able to implement this test by calling get token and having the async callback wait do async sleep

Reply:
All providers are async compatible, it's just that if they don't have their own implementation of _get_token_impl_async, it defaults to the base class one which just calls sync_to_async on it.
Since basic just returns a string it makes sense.

For callback while it technically works, it does seem like missing functionality.
So I added the option for an async callback, and changed the test like you suggested, please take a look